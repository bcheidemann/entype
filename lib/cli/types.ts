import { Lang } from "../emit/types.ts";

export type HelpConfig = {
  help: true;
  invalidArgs: boolean;
  message?: string;
};

export type LangConfig = {
  lang: Lang;
  files: string[];
};

export type Config =
  | HelpConfig
  | LangConfig;
