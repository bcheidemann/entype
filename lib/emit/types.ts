import { Type } from "../types.ts";

export type Lang = "rust" | "typescript";
export type EmitRawFn = (data: string) => void;
export type EmitFn = (
  context: Context,
  name: string,
  type: Type,
  preferInline: boolean,
  code: string,
) => void;
export type EmitAnnotationFn = (annotation: string) => void;
export type GetTypeNameFn = (name?: string) => string;
export type EmitTypeFn = (
  context: Context,
  name: string,
  type: Type,
  preferInline: boolean,
) => string | void;
export type Context = {
  lang: Lang;
  emit: EmitFn;
  emitAnnotation: EmitAnnotationFn;
  emitType: EmitTypeFn;
  getTypeName: GetTypeNameFn;
};
export type EmitterModule = {
  emitType: EmitTypeFn;
};
