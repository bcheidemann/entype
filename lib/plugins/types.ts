import { Context } from "../emit/types.ts";
import { Type } from "../types.ts";

export type PluginName = string;

export type PreEmitHook = (
  context: Context,
  name: string,
  type: Type,
  preferInline: boolean,
  code: string,
) => string | void;

export type Plugin = {
  hooks: {
    preEmitHook?: PreEmitHook;
  };
};
