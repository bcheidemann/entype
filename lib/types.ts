export type Json =
  | string
  | number
  | boolean
  | null
  | JsonArray
  | JsonObject;
export type JsonArray = Json[];
export type JsonObject = { [key: string]: Json };

export type Type =
  | StructType
  | UnionType
  | MapType
  | OptionType
  | ArrayType
  | PrimitiveType
  | NullType
  | UnknownType;
export type StructType = {
  kind: "struct";
  fields: Map<string, Type>;
};
export type UnionType = {
  kind: "union";
  variants: Map<Type["kind"] | PrimitiveType["name"], Type>;
};
export type MapType = {
  kind: "map";
  valueType: Type;
};
export type OptionType = {
  kind: "option";
  valueType: Type;
};
export type ArrayType = {
  kind: "array";
  elementType: Type;
};
export type PrimitiveType = {
  kind: "primitive";
  name: "string" | "number" | "boolean";
};
export type NullType = {
  kind: "null";
};
export type UnknownType = {
  kind: "unknown";
};
export type HomogeneousTypeArray =
  | StructType[]
  | UnionType[]
  | MapType[]
  | OptionType[]
  | ArrayType[]
  | PrimitiveType[]
  | NullType[]
  | UnknownType[];
