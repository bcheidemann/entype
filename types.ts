export type Json =
  | string
  | number
  | boolean
  | null
  | JsonArray
  | JsonObject;
export type JsonArray = Json[];
export type JsonObject = { [key: string]: Json };

export type Type =
  | StructType
  | EnumType
  | MapType
  | OptionType
  | ArrayType
  | PrimitiveType
  | NullType
  | UnknownType;
export type StructType = {
  kind: "struct",
  fields: Map<string, Type>,
};
export type EnumType = {
  kind: "enum",
  variants: Map<string, Type>,
};
export type MapType = {
  kind: "map",
  valueType: Type,
};
export type OptionType = {
  kind: "option",
  valueType: Type,
};
export type ArrayType = {
  kind: "array",
  elementType: Type,
};
export type PrimitiveType = {
  kind: "primitive",
  name: "string" | "number" | "boolean",
};
export type NullType = {
  kind: "null",
};
export type UnknownType = {
  kind: "unknown",
};
export type HomogeneousTypeArray =
  | StructType[]
  | EnumType[]
  | MapType[]
  | OptionType[]
  | ArrayType[]
  | PrimitiveType[]
  | NullType[]
  | UnknownType[];
