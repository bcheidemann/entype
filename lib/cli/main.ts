import "npm:@total-typescript/ts-reset";
import { Json } from "../types.ts";
import { parseJson } from "../parse.ts";
import { collapseTypes } from "../collapse-types.ts";
import { emitRootType } from "../emit/mod.ts";
import { isPromiseFulfilledResult, isPromiseRejectedResult } from "./util.ts";
import { loadPlugin } from "../plugins/mod.ts";
import { getVersion } from "../version.ts";
import { parseArgs } from "./parse-args.ts";

export async function main(
  runtime: "deno" | "node",
  args: string[],
): Promise<0 | 1> {
  const config = parseArgs(args);

  if ("help" in config) {
    if (runtime === "deno") {
      console.log(
        "Usage: entype --lang <rust|typescript> [files]\n",
      );
    } else {
      console.log(
        "Usage: node typegen-json --lang <rust|typescript> [files]\n",
      );
    }

    if ("message" in config) {
      console.error(config.message);
    }

    if (config.invalidArgs) {
      return 1;
    }

    return 0;
  }

  if ("version" in config) {
    console.log(getVersion());
    return 0;
  }

  if ("lang" in config) {
    let exitStatusCode: 0 | 1 = 0;

    const types = await Promise.allSettled(
      config.files.map(async (file) => {
        const json = await Deno.readTextFile(file);
        const obj = JSON.parse(json) as Json;
        return parseJson(obj);
      }),
    );

    if (types.some(isPromiseRejectedResult)) {
      console.error([
        "Failed to parse some files:",
        ...types.map((result, i) => {
          if (result.status === "fulfilled") {
            return `  ${config.files[i]}: OK`;
          } else {
            return `  ${config.files[i]}: ${result.reason}`;
          }
        }),
      ].join("\n"));
      exitStatusCode = 1;
    }

    const type = collapseTypes(
      types
        .filter(isPromiseFulfilledResult)
        .map((result) => result.value),
    );

    const plugins = await Promise.all(config.plugins.map(loadPlugin));

    await emitRootType(config.lang, type, plugins, console.log);

    return exitStatusCode;
  }

  throw new Error("Unreachable (received invalid config)");
}
