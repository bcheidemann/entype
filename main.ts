import "npm:@total-typescript/ts-reset";
import { ArrayType, EnumType, HomogeneousTypeArray, Json, JsonArray, JsonObject, MapType, NullType, OptionType, PrimitiveType, StructType, Type, UnknownType } from "./types.ts";

export function parseJson(obj: Json): Type {
  switch (typeof obj) {
    case "object":
      if (obj === null) {
        return parseNull(obj);
      }
      if (Array.isArray(obj)) {
        return parseJsonArray(obj);
      }
      return parseJsonObject(obj);
    case "string":
      return { kind: "primitive", name: "string" };
    case "number":
      return { kind: "primitive", name: "number" };
    case "boolean":
      return { kind: "primitive", name: "boolean" };
    default:
      throw new Error("Object is not valid JSON");
  }
}

export function parseNull(_obj: null): Type {
  return { kind: "null" };
}

export function parseJsonArray(obj: JsonArray): Type {
  const elementTypes = obj.map(parseJson);
  const elementType = collapseTypes(elementTypes);
  return { kind: "array", elementType };
}

export function parseJsonObject(obj: JsonObject): Type {
  const fields = new Map(
    Object
      .entries(obj)
      .map(([key, value]) => [key, parseJson(value)])
  );
  // TODO: Handle map types (i.e. when keys are not valid identifiers)
  return { kind: "struct", fields };
}

export function isHomogeneousTypeArray(types: Type[]): types is HomogeneousTypeArray {
  return types.every((type) => type.kind === types[0].kind);
}

export function collapseTypes(types: Type[]): Type {
  if (types.length === 0) {
    return { kind: "unknown" };
  }
  return isHomogeneousTypeArray(types)
    ? collapseHomogeneousTypes(types)
    : collapseNonHomogeneousTypes(types);
}

export function collapseHomogeneousTypes(types: HomogeneousTypeArray): Type {
  if (types.length === 0) {
    return { kind: "unknown" };
  }
  switch (types[0].kind) {
    case "struct":
      return collapseStructTypes(types as StructType[]);
    case "enum":
      return collapseEnumTypes(types as EnumType[]);
    case "map":
      return collapseMapTypes(types as MapType[]);
    case "option":
      return collapseOptionTypes(types as OptionType[]);
    case "array":
      return collapseArrayTypes(types as ArrayType[]);
    case "primitive":
      return collapsePrimitiveTypes(types as PrimitiveType[]);
    case "null":
      return collapseNullTypes(types as NullType[]);
    case "unknown":
      return collapseUnknownTypes(types as UnknownType[]);
  }
}

export function collapseNonHomogeneousTypes(types: Type[]): Type {
  const uniqueTypes = new Set<Type["kind"]>(
    types.map((type) => type.kind)
  );

  if (uniqueTypes.has("enum")) {
    return collapseNonHomogeneousTypes(
      types.flatMap((type) => {
        switch (type.kind) {
          case "enum":
            return Array.from(type.variants.values());
          default:
            return type;
        }
      })
    );
  }

  if (uniqueTypes.has("null")) {
    const nonNullTypes = types.filter((type) => type.kind !== "null");
    const collapsedType = collapseTypes(nonNullTypes);
    return { kind: "option", valueType: collapsedType };
  }

  const collectedVariants: {
    struct: StructType[];
    enum: EnumType[];
    map: MapType[];
    option: OptionType[];
    array: ArrayType[];
    primitive: PrimitiveType[];
    null: NullType[];
    unknown: UnknownType[];
  } = {
    struct: [],
    enum: [],
    map: [],
    option: [],
    array: [],
    primitive: [],
    null: [],
    unknown: [],
  };
  for (const type of types) {
    switch (type.kind) {
      case "struct":
        collectedVariants.struct.push(type);
        break;
      case "enum":
        collectedVariants.enum.push(type);
        break;
      case "map":
        collectedVariants.map.push(type);
        break;
      case "option":
        collectedVariants.option.push(type);
        break;
      case "array":
        collectedVariants.array.push(type);
        break;
      case "primitive":
        collectedVariants.primitive.push(type);
        break;
      case "null":
        collectedVariants.null.push(type);
        break;
      case "unknown":
        collectedVariants.unknown.push(type);
        break;
    }
  }
  const variants = new Map(
    Object.entries(collectedVariants)
      .filter(([_variant, types]) => types.length > 0)
      .map(([variant, types]) => [variant, collapseHomogeneousTypes(types)])
  );
  return { kind: "enum", variants };
}

export function collapseStructTypes(types: StructType[]): StructType {
  const keys = new Set<string>();
  for (const type of types) {
    for (const key of type.fields.keys()) {
      keys.add(key);
    }
  }
  const fields = new Map<string, Type>();
  for (const key of keys) {
    const fieldType = collapseTypes(
      types.map((type) => type.fields.get(key) ?? { kind: "null" })
    );
    fields.set(key, fieldType);
  }
  return { kind: "struct", fields };
}

export function collapseEnumTypes(types: EnumType[]): EnumType {
  const variantNames = new Set<string>();
  for (const type of types) {
    for (const variantName of type.variants.keys()) {
      variantNames.add(variantName);
    }
  }
  const variants = new Map<string, Type>();
  for (const variantName of variantNames) {
    const variantType = collapseTypes(
      types
        .map((type) => type.variants.get(variantName))
        .filter(Boolean)
    );
    variants.set(variantName, variantType);
  }
  return { kind: "enum", variants };
}

export function collapseMapTypes(types: MapType[]): MapType {
  const valueTypes = types.map((type) => type.valueType);
  const valueType = collapseTypes(valueTypes);
  return { kind: "map", valueType };
}

export function collapseOptionTypes(types: OptionType[]): OptionType {
  const valueTypes = types.map((type) => type.valueType);
  const valueType = collapseTypes(valueTypes);
  return { kind: "option", valueType };
}

export function collapseArrayTypes(types: ArrayType[]): ArrayType {
  const elementTypes = types.map((type) => type.elementType);
  const elementType = collapseTypes(elementTypes);
  return { kind: "array", elementType };
}

export function collapsePrimitiveTypes(types: PrimitiveType[]): PrimitiveType | EnumType {
  const names = new Set<PrimitiveType["name"]>();
  for (const type of types) {
    names.add(type.name);
  }
  if (names.size === 1) {
    return { kind: "primitive", name: names.values().next().value };
  }
  const variants = new Map(
    Array
      .from(names.values())
      .map((name) => [name, { kind: "primitive", name }] as const)
  );
  return { kind: "enum", variants };
}

export function collapseNullTypes(_types: NullType[]): NullType {
  return { kind: "null" };
}

export function collapseUnknownTypes(_types: UnknownType[]): UnknownType {
  return { kind: "unknown" };
}

if (import.meta.main) {
  console.log('Ok');
}
