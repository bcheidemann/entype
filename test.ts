import {
  describe,
  it,
} from "https://deno.land/std@0.198.0/testing/bdd.ts";
import { assertSnapshot } from "https://deno.land/std@0.198.0/testing/snapshot.ts";
import { parseJson } from "./main.ts";
import { Json } from "./types.ts";

describe("parseJson", () => {
  async function run(
    ctx: Deno.TestContext,
    filename: string,
  ) {
    const json = await Deno.readTextFile(`./fixtures/${filename}`);
    const obj = JSON.parse(json) as Json;
    const type = parseJson(obj);
    await assertSnapshot(ctx, type);
  }

  it("parses array-0.json", async (ctx) => {
    await run(ctx, "array-0.json");
  });

  it("parses array-1.json", async (ctx) => {
    await run(ctx, "array-1.json");
  });

  it("parses array-2.json", async (ctx) => {
    await run(ctx, "array-2.json");
  });

  it("parses boolean-false.json", async (ctx) => {
    await run(ctx, "boolean-false.json");
  });

  it("parses boolean-true.json", async (ctx) => {
    await run(ctx, "boolean-true.json");
  });

  it("parses map-0.json", async (ctx) => {
    await run(ctx, "map-0.json");
  });

  it("parses map-1.json", async (ctx) => {
    await run(ctx, "map-1.json");
  });

  it("parses map-2.json", async (ctx) => {
    await run(ctx, "map-2.json");
  });

  it("parses null.json", async (ctx) => {
    await run(ctx, "null.json");
  });

  it("parses number-negative.json", async (ctx) => {
    await run(ctx, "number-negative.json");
  });

  it("parses number-positive.json", async (ctx) => {
    await run(ctx, "number-positive.json");
  });

  it("parses number-zero.json", async (ctx) => {
    await run(ctx, "number-zero.json");
  });

  it("parses object-0.json", async (ctx) => {
    await run(ctx, "object-0.json");
  });

  it("parses object-1.json", async (ctx) => {
    await run(ctx, "object-1.json");
  });

  it("parses object-2.json", async (ctx) => {
    await run(ctx, "object-2.json");
  });

  it("parses object-3.json", async (ctx) => {
    await run(ctx, "object-3.json");
  });

  it("parses object-4.json", async (ctx) => {
    await run(ctx, "object-4.json");
  });

  it("parses object-5.json", async (ctx) => {
    await run(ctx, "object-5.json");
  });

  it("parses object-6.json", async (ctx) => {
    await run(ctx, "object-6.json");
  });

  it("parses object-7.json", async (ctx) => {
    await run(ctx, "object-7.json");
  });

  it("parses object-8.json", async (ctx) => {
    await run(ctx, "object-8.json");
  });
});
