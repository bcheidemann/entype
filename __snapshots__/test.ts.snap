export const snapshot = {};

snapshot[`parseJson > parses array-0.json 1`] = `
{
  elementType: {
    kind: "primitive",
    name: "number",
  },
  kind: "array",
}
`;

snapshot[`parseJson > parses array-1.json 1`] = `
{
  elementType: {
    fields: Map(2) {
      "key1" => {
        kind: "primitive",
        name: "string",
      },
      "key2" => {
        kind: "primitive",
        name: "string",
      },
    },
    kind: "struct",
  },
  kind: "array",
}
`;

snapshot[`parseJson > parses array-2.json 1`] = `
{
  elementType: {
    fields: Map(3) {
      "key1" => {
        kind: "primitive",
        name: "string",
      },
      "key2" => {
        kind: "primitive",
        name: "string",
      },
      "optional" => {
        kind: "option",
        valueType: {
          kind: "primitive",
          name: "string",
        },
      },
    },
    kind: "struct",
  },
  kind: "array",
}
`;

snapshot[`parseJson > parses boolean-false.json 1`] = `
{
  kind: "primitive",
  name: "boolean",
}
`;

snapshot[`parseJson > parses boolean-true.json 1`] = `
{
  kind: "primitive",
  name: "boolean",
}
`;

snapshot[`parseJson > parses map-0.json 1`] = `
{
  fields: Map(4) {
    "a.b.c" => {
      kind: "primitive",
      name: "number",
    },
    "a.b.d" => {
      kind: "primitive",
      name: "number",
    },
    "a.b.e" => {
      kind: "primitive",
      name: "number",
    },
    "a.b.f" => {
      kind: "primitive",
      name: "number",
    },
  },
  kind: "struct",
}
`;

snapshot[`parseJson > parses map-1.json 1`] = `
{
  fields: Map(4) {
    "1" => {
      kind: "primitive",
      name: "number",
    },
    "2" => {
      kind: "primitive",
      name: "number",
    },
    "3" => {
      kind: "primitive",
      name: "number",
    },
    "4" => {
      kind: "primitive",
      name: "number",
    },
  },
  kind: "struct",
}
`;

snapshot[`parseJson > parses map-2.json 1`] = `
{
  fields: Map(2) {
    "key.2" => {
      kind: "primitive",
      name: "number",
    },
    "key1" => {
      kind: "primitive",
      name: "number",
    },
  },
  kind: "struct",
}
`;

snapshot[`parseJson > parses null.json 1`] = `
{
  kind: "null",
}
`;

snapshot[`parseJson > parses number-negative.json 1`] = `
{
  kind: "primitive",
  name: "number",
}
`;

snapshot[`parseJson > parses number-positive.json 1`] = `
{
  kind: "primitive",
  name: "number",
}
`;

snapshot[`parseJson > parses number-zero.json 1`] = `
{
  kind: "primitive",
  name: "number",
}
`;

snapshot[`parseJson > parses object-0.json 1`] = `
{
  fields: Map(1) {
    "key" => {
      kind: "primitive",
      name: "string",
    },
  },
  kind: "struct",
}
`;

snapshot[`parseJson > parses object-1.json 1`] = `
{
  fields: Map(1) {
    "key" => {
      kind: "primitive",
      name: "number",
    },
  },
  kind: "struct",
}
`;

snapshot[`parseJson > parses object-2.json 1`] = `
{
  fields: Map(1) {
    "key" => {
      kind: "primitive",
      name: "boolean",
    },
  },
  kind: "struct",
}
`;

snapshot[`parseJson > parses object-3.json 1`] = `
{
  fields: Map(1) {
    "key" => {
      elementType: {
        kind: "primitive",
        name: "string",
      },
      kind: "array",
    },
  },
  kind: "struct",
}
`;

snapshot[`parseJson > parses object-4.json 1`] = `
{
  fields: Map(1) {
    "key" => {
      fields: Map(1) {
        "key" => {
          kind: "primitive",
          name: "string",
        },
      },
      kind: "struct",
    },
  },
  kind: "struct",
}
`;

snapshot[`parseJson > parses object-5.json 1`] = `
{
  fields: Map(1) {
    "key" => {
      kind: "null",
    },
  },
  kind: "struct",
}
`;

snapshot[`parseJson > parses object-6.json 1`] = `
{
  fields: Map(1) {
    "key" => {
      elementType: {
        fields: Map(1) {
          "key" => {
            kind: "primitive",
            name: "string",
          },
        },
        kind: "struct",
      },
      kind: "array",
    },
  },
  kind: "struct",
}
`;

snapshot[`parseJson > parses object-7.json 1`] = `
{
  fields: Map(1) {
    "key" => {
      elementType: {
        kind: "enum",
        variants: Map(2) {
          "boolean" => {
            kind: "primitive",
            name: "boolean",
          },
          "string" => {
            kind: "primitive",
            name: "string",
          },
        },
      },
      kind: "array",
    },
  },
  kind: "struct",
}
`;

snapshot[`parseJson > parses object-8.json 1`] = `
{
  fields: Map(8) {
    "key1" => {
      kind: "primitive",
      name: "boolean",
    },
    "key2" => {
      kind: "primitive",
      name: "boolean",
    },
    "key3" => {
      kind: "null",
    },
    "key4" => {
      kind: "primitive",
      name: "number",
    },
    "key5" => {
      kind: "primitive",
      name: "number",
    },
    "key6" => {
      kind: "primitive",
      name: "string",
    },
    "key7" => {
      elementType: {
        kind: "primitive",
        name: "number",
      },
      kind: "array",
    },
    "key8" => {
      fields: Map(1) {
        "key1" => {
          kind: "primitive",
          name: "string",
        },
      },
      kind: "struct",
    },
  },
  kind: "struct",
}
`;
